### PR TITLE
1.13: Readd Anti-Xray

### LINKED ISSUE #1183: Obfuscated ores showing when an NPC builder excavates down to stone before building something
### What behaviour is expected:
The actual ores and stones etc that were on the ground

### What behaviour is observed:
Obfuscated ores(the ones that show when players attempt to use an xray mod when anti-xray is switched on in paper.yml) are showing when the builder digs down to a cetain level when attmepting to build something, it's nothing major but just something I thought I'd point out and to ask if there's a way around it without having to remove the anti-xray feature. Also the obfuscated ores do dissapear when a player attempts to mine or hit them. EDIT - The obfuscated ores also seem to be acting as a light source.

### Steps/models to reproduce:
Get a Citizens NPC with builder trait to build something where digging down is required

### Plugin list:
Plugins (125): ActionHealth, AdvancedPortals, AgarthaLib, AlwaysOnline, AntiPortalTrapping, AsyncWorldEdit, AsyncWorldEditInjector, AutoUpdaterAPI, BetterDrops, Blacksmith, BlockStatus, BottleRecycling, Broadcaster, Builder, ChangeSkin, chatemotes, Citizens, CitizensCMD, CityWorld, ClickSort, ColorCodes, ColoredAnvils, CombatTagPlus, CommandPrompter, Craftconomy3, Crash, CrispyAntiWDL, CS-CoreLib, CustomJoinItems, CustomSoundManagerAPI, DamageIndicator, DeluxeChat, DeluxeMenus, Denizen, DiceFurniture, DiscordSRV, DragonEggRespawn, DynamicSigns, EazyEmojis-Lite, EditableSign, Essentials, EssentialsChat, EssentialsGeoIP, eZProtector, FirstJoinPlus, FOLFriends, ForceResourcepacks, ForgeModBlocker, FurnitureLib, gravy, HolographicDisplays, icJukeBox, InfoBook, InventoryFullPlus, InventoryOverflow, InventoryTweaks, LastHolo, LibsDisguises, LuckPerms, MaintenanceSpigot, Mentions.gg, Missing_Recipes, MoneyDrop, MultiInv, Multiverse-Core, Multiverse-Portals, MVdWPlaceholderAPI, MVdWUpdater, NoPos, numberKeyBind, OnlinePlayersMenu, OpenInv, PacketListenerApi, ParadiseLand, PlaceholderAPI, PluginManagerReloaded, ProtectionLib, Protectiontime, ProtocolLib, PulseHealth, Questioner, Realistic-Swimming, RealisticLiquid, RealisticMC, SavageDeathCompass, SecurityVillagers, SelectionVisualizer, Sentinel, ShulkerBoxPlus, SilkSpawners, SimpleSit, Skript, Slime_Chunk, StackableItems, StatusBlock, SuperTrails, SurvivalMechanics, SurvivalMechanicsFlagAddon, SwordDefence, TabList, TintHealth, TitleManager, TokenEnchant, Towny, TownyChat, TownyGUI, TownyNameUpdater, TownyTitle, TradeMe, Trails, UltimateDrugs, UltimateTrash, UltimateTroll, Vault, ViaBackwards, ViaRewind, ViaVersion, Votifier, VotingPlugin, WGRegionEvents, WGTitles, Wild, WorldEdit, WorldGuard, WorldGuardExtraFlags

### Paper build number:
This server is running Paper version git-Paper-1470 (MC: 1.12.2) (Implementing API version 1.12.2-R0.1-SNAPSHOT)

### Anything else:
Refer to this link please

https://github.com/mcmonkey4eva/Builder/issues/3#issuecomment-401529455

### LINKED ISSUE #1230: 1.13: Readd Anti Xray
Patch removed, needs to be re added. 
0235-Anti-Xray.patch

### PR DESCRIPTION
**Pull request by stonar96**
Resolves #1230 
Probably resolves #1183 

Special thanks to @Spottedleaf who helped me with decompile errors. Thank you for the support on IRC and thanks to everyone who motivated me by being interested in this.

https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=M9XHJ5E4XNXGE

### CODE REVIEW

**Comment by aikar** [comment]
Can you undo the submodule updates, as well as please update your git and rebuild?

Older gits cause extra patch diffs that shouldn't of been changed.

**Comment by stonar96** [comment]
Done. Updated to git version 2.18.0.windows.1. `./paper patch` and `./paper rebuild` changed nothing.

**Comment by stonar96** [comment]
Fixed.

**Comment by NullCase** [comment]
hi @stonar96 

It looks like you added your paypal link above, which is great.

What should I watch for to know this has been added to Paper 1.13?

**Comment by stonar96** [comment]
@NullCase You will see it here in this PR when it says something like "... merged commit ... into ..." and/or the PR gets closed.

File: Spigot-Server-Patches/0348-Anti-Xray.patch
Comment (aikar) [inline Spigot-Server-Patches/0348-Anti-Xray.patch:703]: we can do world.getTypeIfLoaded here instead and save a map lookup

**Review by aikar — APPROVED** [review]
Just 1 thing is all i see as a performance improvement, but i'm not really holding it up for that.

just looking for more peers to sign off first too before pulling.

but if you can change this isLoaded/getType to a getTypeIfLoaded, nicer on performance.

File: Spigot-Server-Patches/0348-Anti-Xray.patch
Comment (stonar96) [reply Spigot-Server-Patches/0348-Anti-Xray.patch:703]: Will do.

**Review by zachbr — APPROVED** [review]
Traditionally OBFHELPERs go in the patch that they are used in, not the utils patch. But I suppose it doesn't really matter, and it's nothing to hold this up over.

**Comment by stonar96** [comment]
Thanks for reviewing. @zachbr I can change it. My thought was to keep the Anti-Xray patch a bit cleaner. So I only added functionality to this patch.

**Comment by zachbr** [comment]
I'm not super worried about it. I don't know how the rest of core team feels but w/e. If they get moved, that's fine. If not, also good. :+1: 

The only downside is that the patch isn't portable without them, you'd need a piece of the utils patch.

**Comment by DenWav** [comment]
What if there was just an OBFHELPER patch that contained all OBFHELPERs early in the patch process (like patch `#4` or whatever)? All patches afterwards use OBFHELPERs anywhere possible. Would keep all of these similar changes together in one patch and might help reduce the burden of upgrading. The OBFHELPER patch itself might be annoying, but once it's done any patch after that point that uses OBFHELPERs might require less modification for upgrades. OBFHELPERs usually _aren't_ directly related to the patch at hand other than simply being called, I think it makes sense to group them together.

This isn't really the right place to have this discussion either, so uh, I say it's fine.

**Comment by aikar** [comment]
Obf helpers should be easy to maintain. Ones that are likely to be used multiple times are nice to be in an early patch, but leaving fringe ones in the using patch so they can be dropped if the patch is dropped.

However this patch is one we know we want to keep forever so I'm fine with leaving them in mcutil. 

I'm also OK with using mcutil as the home for generic helpers because the helpers are a utility concept.

**Comment by zachbr** [comment]
As always thank you @stonar96, and @Spottedleaf as well for lending a hand with the decomp fixes.

Builds 170+

**Comment by NullCase** [comment]
@stonar96 

I've started testing this. Going to let my players try and break it for a day and see how things go.

It seems like frame rates take a performance hit so i've created an issue. https://github.com/PaperMC/Paper/issues/1376

**Pull request merged**
